Implement strongly typed layers

RmStorm opened the pull request:
This PR adds most layer methods defined here:
https://docs.mapbox.com/mapbox-gl-js/api/map/#instance-members-layers

And it adds types for all the layers, paint properties and layout properties defined here:
https://docs.mapbox.com/style-spec/reference/layers/

---

Hello hello, [I'm back](https://github.com/yukinarit/mapbox-gl-rs/pull/62) with a huge PR this time. I started using this crate to prototype some shader stuff and I got completely carried away honestly.. So uhm.. I know it's poor form to launch giant PR's like this one at an open source project :sweat_smile: very sorry about that!

I'm putting this PR in draft mode because I think we need to discuss a bunch of things if you are to consider merging it. I hope you can find some time to engage! ~~I cleaned up the work here just a little bit and divided it into 3 commits.~~

~~- One that is pure admin crap to get examples running (deleting and changing Cargo.toml's and Trunk.toml files).~~
~~- Than a commit with a complete refactoring of the layer's and a bunch of other changes sprinkled in, if we actually get to the point where we think the work is worth merging I will obviously clean this up much more.~~
~~- And the final commit contains example updates.~~
This PR is now gated behind: https://github.com/yukinarit/mapbox-gl-rs/pull/65 that basically gets all the examples running again some of which were broken. It also fixes other minor issues I ran into along the way.

The main attraction is basically support for custom webgl layers and strongly typed layout and paint objects. Take a quick look at this screen cap for the coolness of the strong typing! autocomplete!

![demo](https://github.com/yukinarit/mapbox-gl-rs/assets/13096654/7de4dc75-9206-4596-bfcb-055ae7baae90)

yukinarit commented:
Hi @RmStorm!

Thank you so much for PRs! I am a bit slow but let me review little by little :pray:

RmStorm commented:
> Hi @RmStorm!
> 
> Thank you so much for PRs! I am a bit slow but let me review little by little 🙏

No problem at all, my pleasure. And there is no hurry on the review :sweat_smile:. I do think we should merge https://github.com/yukinarit/mapbox-gl-rs/pull/65 first. I couldn't get most examples to run due to Trunk changes and when making all these changes to layers I also ran into other small issues such as `pub fn getCenter(this: &Map) -> LngLat;` being missing. I moved that into the other  https://github.com/yukinarit/mapbox-gl-rs/pull/65 as well. So if we merge that. This one will be much easier to review.

yukinarit commented:
sorry, I did "rebase and merge" for #64, which altered the commit hash of your commit. Please rebase this branch to main :pray:

RmStorm commented:
Rebased and cleaned up a little. I've added about [half the layer methods on map](https://docs.mapbox.com/mapbox-gl-js/api/map/#instance-members-layers). I've also added about [half the layers](https://docs.mapbox.com/style-spec/reference/layers/), namely: `Background`, `Fill`, `Line`, `Symbol`, `Circle` and `Custom`.

It's not _that_ much work to finish it, e.g. add all the layer methods and implement all the layers. But we need to figure out if the choices I've made or sane! And if we have to change things we should do that before I implement the rest. So I'm going to make a bunch of comments in the code that will hopefully make reviewing easier and I'm going to call out the things where I think we might want to change things.

RmStorm commented on src/layer.rs:
```diff
@@ -1,102 +1,732 @@
+use serde::{Deserialize, Serialize};
+use wasm_bindgen::{closure::Closure, JsValue};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Expression<T> {
+    Enum(T),
+    String(String),
+    Number(f64),
+    Bool(bool),
+    List(Vec<Expression<T>>),
+}
```
This has basically replaced:
```rust
/// Layout property can be either value in String or Number e.g. 0.25
/// or tuple of 2 elements, ("get", "icon").
#[derive(Debug, Clone, Serialize)]
#[serde(untagged)]
pub enum LayoutProperty {
    String(String),
    Number(f64),
    FeatureProperty((String, String)),
}
```

Since all of these 'feature properties' can actually be full blown [mapbox expressions](https://docs.mapbox.com/style-spec/reference/expressions/). I decided to lean in and just kinda add types for a [Lisp](https://stopa.io/post/222) since that's what the expressions are! I considered trying to make these in such a way that the expression always evaluates to a certain type but realized that I wasn't really smart enough hahaha. Either that or it's not really possible, I didn't give that too much thought.

But then when I got further I came across a lot of properties such as [line-join](https://docs.mapbox.com/style-spec/reference/layers/#layout-line-line-join) that are actually enums. So I did try to pull those into the expressions but it came out a little bit wonky I think.. With the `Enum` inside the `enum`. Maybe they should just be strings after all? Anyway, please have a look and a think about it!

Irrespective of where we take the rest of the layer stuff, having a good representation of the mapbox expressions is quite nice so it makes sense to spend some time to find an approach that makes us happy!

RmStorm commented on src/layer.rs:
```diff
@@ -1,102 +1,732 @@
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GetLayer {
```
This is kinda sad but the return value of [map.getLayer()](https://docs.mapbox.com/mapbox-gl-js/api/map/#map#getlayer) is not really a normal layer such as you would get inside [map.getStyle()](https://docs.mapbox.com/mapbox-gl-js/api/map/#map#getstyle). Instead it is it's own wonky thing that is not really parsable.. So I made this getLayer struct in such a way that the two mapbox examples that use `getLayer` are implementable but that's it.. There is a [bigger comment about it here](https://github.com/yukinarit/mapbox-gl-rs/pull/64/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R902-R909) as well.

RmStorm commented on src/layer.rs:
```diff
@@ -1,102 +1,732 @@
+    pub visibility: Option<Visibility>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum Layer {
+    Custom(CustomLayer),
+    Fill(FillLayer),
+    Background(BackgroundLayer),
+    Line(LineLayer),
+    Symbol(SymbolLayer),
+    Circle(CircleLayer),
+}
+
+pub trait IntoLayer {
+    fn into_layer(self) -> Layer;
+}
```
Now we get to the heart of the matter. A strongly typed implementation of the layers. Every layer implement the `IntoLayer` trait so that these layers can all be passed into `map.add_layer()`. By using the serde annotations we can actually deserialize the output of `getStyle` and get out all correct strongly types layers.

It also allows for individual `Paint` and `Layout` properties. That is really exciting because it means you get autocomplete on all of those features! Which is super nice in my opinion. I'm just confused about why they haven't done it like that in the Javascript/Typesciprt implementation and I'm wondering if I'm really missing something that invalidates this approach? Let me know what you think.

RmStorm commented on src/layer.rs:
```diff
@@ -1,102 +1,732 @@
+    pub inner: LayerBase,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub rendering_mode: Option<String>,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub on_add: js_sys::Function,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub on_remove: js_sys::Function,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub prerender: js_sys::Function,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub render: js_sys::Function,
```
I think there might be other ways of moving these closures through the wasm boundary but this was the first one I figured out. If you know of a better approach I'm all ears. This is my first real foray into wasm-bindgen land so idk. I'm just winging it.

yukinarit replied:
hmm I don't know the best practice ..

RmStorm commented on src/layer.rs:
```diff
@@ -1,102 +1,732 @@
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FillLayer {
+    #[serde(flatten)]
+    pub inner: LayerBase,
+    #[serde(default)]
+    pub layout: FillLayout,
+    #[serde(default)]
+    pub paint: FillPaint,
+}
```
This is _the_ typical layer. It has a layer base with stuff like, source, max and min zoom etc. all flattened directly into the layer. And it has strongly typed layout and paint properties that default to basically all `None`'s quite readable I think and very usable in combination with autocomplete!

The only downside is that setting the source for example is `my_layer.inner.source = …` instead of `my_layer.source = …` what do you think?

RmStorm replied:
Yeah I actually agree. I will change that.

RmStorm commented on src/lib.rs:
```diff
@@ -666,6 +676,51 @@ pub struct CameraAnimationOptions {
+    }
+}
+
+impl From<serde_wasm_bindgen::Error> for SerdePathFindingError {
+    fn from(error: serde_wasm_bindgen::Error) -> Self {
+        let message = format!("{}", error);
+        let search_term = "::<impl serde::de::Deserialize for ";
+        let js_error: JsValue = error.into();
+        let mut path = vec![];
+        let mut raw_error = message.clone();
+
+        if let Some(e) = js_error.dyn_ref::<js_sys::Error>() {
+            raw_error = format!("{:?}", e);
+            for line in raw_error.lines() {
+                if let Some(start) = line.find(search_term) {
+                    let rest = &line[start + search_term.len()..];
+                    if let Some(end) = rest.find('>') {
+                        path.push(rest[..end].into())
+                    }
+                }
+            }
+        }
+        path = path.into_iter().rev().collect();
+        SerdePathFindingError {
+            message,
+            path,
+            raw_error,
+        }
+    }
+}
```
The errors generated by wasm_bindgen where literally unintelligible. I made this super dum wrapper for debugging purposes and will delete it before we merge the PR :sweat_smile:.

RmStorm commented on src/style.rs:
```diff
@@ -14,3 +16,49 @@ impl StyleOptions {
+    pub pitch: Option<f64>,
+    pub sprite: Option<String>,
+    pub glyphs: Option<String>,
+    pub layers: Option<Vec<Layer>>,
```
This is super cool. It just works :tm:, you can call `map.get_style()` and get out fully strongly typed layers.

DennisJensen95 commented:
Good work @RmStorm! I am also gonna need it in my setup, thanks for the work 🚀

yukinarit commented on src/layer.rs:
```diff
@@ -1,102 +1,732 @@
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub layout: Option<Layout>,
+    #[serde(rename = "type")]
+    pub type_: String,
```
I wonder which is better :thinking: 

```rust
#[serde(rename = "type")]
pub type_: String
```
vs.
```rust
pub r#type: String
```

yukinarit commented on src/layer.rs:
```diff
@@ -1,102 +1,732 @@
+impl EnumMarker for LineJoin {}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
```
Feel like having [skip_serializing_none](https://docs.rs/serde_with/latest/serde_with/attr.skip_serializing_none.html) from serde_with, as all the fields are with `Option::is_none` maybe the refactoring should happen in a separate PR though

yukinarit commented on src/lib.rs:
```diff
@@ -538,6 +541,13 @@ impl Map {
+        //     &JsValue::from("mappy"),
+        //     &inner,
+        // )
+        // .unwrap();
```
Will you remove this?

yukinarit reviewed (commented):
Hi @RmStorm The PR basically looks good to me! Please finish the remaining work, so I can merge and ship :rocket:

RmStorm commented:
Sorry for the radio silence! I'm kind off in the middle of moving so I have very little time right now :sweat_smile: I hope I will find an evening or two sometime soon :crossed_fingers: I'm still very much intending to wrap this up!

yukinarit commented on README.md:
```diff
@@ -55,7 +55,7 @@ trunk serve
 	* [ ] [Sources](https://docs.mapbox.com/mapbox-gl-js/api/map/#instance-members-sources)
 	* [x] [Images](https://docs.mapbox.com/mapbox-gl-js/api/map/#instance-members-images)
-	* [ ] [Layers](https://docs.mapbox.com/mapbox-gl-js/api/map/#instance-members-layers)
+	* [/] [Layers](https://docs.mapbox.com/mapbox-gl-js/api/map/#instance-members-layers)
```
is this a typo?

RmStorm replied:
It misses a bunch of methods. I felt a bit uncomfortably implementing them since they're not used in any examples and there wouldn't be any testing on them either. I figured crossing it off entirely would not be right and left this as a placeholder to myself to decide what to do :sweat_smile: I forgot about it instead. wdyt is the correct thing? Maybe just cross it off?

yukinarit replied:
It already has a good support of layers. Of course it is not perfect, but we can incrementally add features.  

For me it is okay!

yukinarit commented on src/lib.rs:
```diff
@@ -669,7 +671,18 @@ pub struct CameraAnimationOptions {
+    // Note that this is using the `log` function imported above during
+    // `bare_bones`
+    ($($t:tt)*) => (log(&format_args!($($t)*).to_string()))
+}
```
nice!

RmStorm replied:
nice but not actually used :see_no_evil: It was for debugging.

yukinarit reviewed (approved):
LGTM!

RmStorm commented:
> LGTM!

I cannot merge! You have to do it :sweat_smile: but I can write a commit message. I suggest something like:

--------

Implement strongly typed layers

This commit adds most layer methods defined here:
https://docs.mapbox.com/mapbox-gl-js/api/map/#instance-members-layers

And it adds types for all the layers, paint properties and layout properties defined here:
https://docs.mapbox.com/style-spec/reference/layers/

yukinarit commented:
Merging! :rocket:

yukinarit commented:
I put the suggested message in the PR description!

yukinarit commented:
I will bump and release the new version soon :+1:
